### PR TITLE
Optimized arrays in redscat algorithm

### PR DESCRIPTION
**Pull request by Bayyapu**

### CODE REVIEW

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

File: src/mpi/coll/allreduce.c [hunk 1]
```diff
@@ -712,52 +716,41 @@ int MPIR_Allreduce_reduce_scatter_allgather(
...
+      for (i = 0; i < pof2; i++) {
+          cnts[i] = newcount / 2;
+          newcount = newcount - cnts[i];
       }
+      cnts[iter-1] = (newcount / 2) + (newcount % 2);
 
-      disps[0] = 0;
-      for (i=1; i<pof2; i++)
-          disps[i] = disps[i-1] + cnts[i-1];
+      disps[-1] = 0;
```
Comment (akhillanger) [inline src/mpi/coll/allreduce.c]: Why was this changed to disps[-1], that looks incorrect?

File: src/mpi/coll/allreduce.c [hunk 2]
```diff
@@ -712,52 +716,41 @@ int MPIR_Allreduce_reduce_scatter_allgather(
...
+      for (i = 1; i < iter; i++)
+          disps[i] = disps[i-1] + cnts[i];
```
Comment (akhillanger) [inline src/mpi/coll/allreduce.c]: Wasn't cnts[i-1] correct already?

File: src/mpi/coll/allreduce.c [hunk 3]
```diff
@@ -649,8 +649,12 @@ int MPIR_Allreduce_reduce_scatter_allgather(
...
     /* find nearest power-of-two less than or equal to comm_size */
     pof2 = 1;
-    while (pof2 <= comm_size) pof2 <<= 1;
-    pof2 >>=1;
+    if (comm_size == 1)
+        iter = 0;
+    else
+        iter = -1;
+    while (pof2 <= comm_size) { iter ++; pof2 <<= 1;}
+    pof2 >>= 1;
```
Comment (akhillanger) [inline src/mpi/coll/allreduce.c]: This would set iter=1 when comm_size=1, is that intentional? iter should be zero when comm_size=1.

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp

**Comment by Bayyapu** [comment]
test:jenkins/ch3/tcp
